In [6]:
import os

train_path = "/content/train/train"

classes = sorted(os.listdir(train_path))
print("Number of classes:", len(classes))

for cls in classes[:5]:
    print(cls, len(os.listdir(os.path.join(train_path, cls))))

Number of classes: 20
0 900
1 900
10 900
11 900
12 900


In [7]:
import os
import cv2
import numpy as np

data = []
labels = []

train_path = "/content/train/train"

for label in os.listdir(train_path):

    class_path = os.path.join(train_path, label)

    if os.path.isdir(class_path):

        files = os.listdir(class_path)[:100]   # only first 100 images per class

        for file in files:

            img_path = os.path.join(class_path, file)

            img = cv2.imread(img_path)

            if img is not None:
                img = cv2.resize(img, (64,64))
                data.append(img.flatten())
                labels.append(int(label))

X = np.array(data)
y = np.array(labels)

print("X shape:", X.shape)
print("y shape:", y.shape)

X shape: (2000, 12288)
y shape: (2000,)


In [9]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42
)

print("X_train:", X_train.shape)
print("X_test:", X_test.shape)
print("y_train:", y_train.shape)
print("y_test:", y_test.shape)

X_train: (1600, 12288)
X_test: (400, 12288)
y_train: (1600,)
y_test: (400,)


In [10]:
from sklearn.svm import SVC

svm_model = SVC(kernel='linear')

svm_model.fit(X_train, y_train)

print("Training Complete!")

Training Complete!


In [11]:
from sklearn.metrics import accuracy_score

y_pred = svm_model.predict(X_test)

accuracy = accuracy_score(y_test, y_pred)

print("Accuracy:", accuracy)

Accuracy: 0.9975


In [12]:
from sklearn.metrics import classification_report

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        15
           1       1.00      1.00      1.00        22
           2       1.00      1.00      1.00        25
           3       1.00      1.00      1.00        18
           4       1.00      1.00      1.00        21
           5       1.00      1.00      1.00        16
           6       0.97      1.00      0.98        28
           7       1.00      0.94      0.97        18
           8       1.00      1.00      1.00        20
           9       1.00      1.00      1.00        16
          10       1.00      1.00      1.00        24
          11       1.00      1.00      1.00        25
          12       1.00      1.00      1.00        21
          13       1.00      1.00      1.00        17
          14       1.00      1.00      1.00        18
          15       1.00      1.00      1.00        18
          16       1.00      1.00      1.00        19
          17       1.00    

Conclusion

A Support Vector Machine (SVM) classifier was developed for Hand Gesture Recognition using a dataset containing 20 gesture classes. The images were resized to 64×64 pixels and converted into feature vectors using flattened pixel values.
The dataset was split into training and testing sets in an 80:20 ratio. A Linear SVM model was trained on the training data and evaluated on the testing data.
Results
Accuracy: 99.75%
Precision: 99.8%
Recall: 99.8%
F1-Score: 99.8%
The model demonstrated excellent performance in classifying hand gestures, achieving near-perfect accuracy across multiple gesture classes. The results indicate that SVM is highly effective for this hand gesture recognition task when trained on properly preprocessed image data.